File: android/src/main/java/com/alpha0010/pdf/PdfViewerBase.kt
package com.alpha0010.pdf

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import android.util.LruCache
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.common.MapBuilder
import com.facebook.react.uimanager.events.RCTEventEmitter
import kotlinx.coroutines.*
import java.io.File
import java.io.FileNotFoundException
import java.util.concurrent.locks.Lock
import kotlin.concurrent.withLock

// MARK: - Tap Zone Helper

enum class TapZone { LEFT, MIDDLE, RIGHT }

fun classifyTapZone(tapX: Float, viewWidth: Int, edgeTapZone: Float): TapZone {
    val edgeRatio = edgeTapZone / 100f
    val leftEdge = viewWidth * edgeRatio
    val rightEdge = viewWidth * (1f - edgeRatio)
    return when {
        tapX < leftEdge -> TapZone.LEFT
        tapX > rightEdge -> TapZone.RIGHT
        else -> TapZone.MIDDLE
    }
}

// MARK: - Shared Manager Constants

object PdfViewerConstants {
    const val COMMAND_RESET_ZOOM = 1
    const val COMMAND_SCROLL_TO_PAGE = 2
    const val COMMAND_CLEAR_STROKES = 3

    fun commandsMap(): Map<String, Int> = MapBuilder.of(
        "resetZoom", COMMAND_RESET_ZOOM,
        "scrollToPage", COMMAND_SCROLL_TO_PAGE,
        "clearStrokes", COMMAND_CLEAR_STROKES
    )

    fun bubblingEventTypes(): Map<String, Any> = MapBuilder.builder<String, Any>()
        .put("onPdfError", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onPdfError")))
        .put("onPdfLoadComplete", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onPdfLoadComplete")))
        .put("onPageChange", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onPageChange")))
        .put("onZoomChange", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onZoomChange")))
        .put("onTap", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onTap")))
        .put("onMiddleClick", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onMiddleClick")))
        .put("onDrawingStart", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onDrawingStart")))
        .put("onDrawingEnd", MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", "onDrawingEnd")))
        .build()

    /** Shared findViewByTag + getAnnotations for Module classes */
    fun getAnnotationsForTag(context: ReactApplicationContext, viewTag: Int, promise: Promise) {
        context.runOnUiQueueThread {
            try {
                val view = context.currentActivity
                    ?.findViewById<View>(android.R.id.content)
                    ?.let { findViewByTag(it, viewTag) } as? PdfViewerBase

                if (view != null) {
                    promise.resolve(view.getAnnotations())
                } else {
                    promise.reject("E_VIEW_NOT_FOUND", "PDF view not found for tag $viewTag")
                }
            } catch (e: Exception) {
                promise.reject("E_UNKNOWN", e.message)
            }
        }
    }

    private fun findViewByTag(view: View, tag: Int): View? {
        if (view.id == tag) return view
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                val result = findViewByTag(view.getChildAt(i), tag)
                if (result != null) return result
            }
        }
        return null
    }
}

// MARK: - PdfViewerBase

/**
 * Base class for PagingPdfView and ZoomablePdfScrollView.
 * Contains shared props, PDF loading, annotation parsing, drawing controller, stroke management,
 * RN event emission, and DrawingControllerDelegate implementation.
 */
@SuppressLint("ViewConstructor")
abstract class PdfViewerBase(context: Context, protected val pdfMutex: Lock) : FrameLayout(context), DrawingControllerDelegate {

    // MARK: - Props

    protected var mSource = ""
    protected var mAnnotations = ""
    protected var mParsedAnnotations: List<AnnotationPage> = emptyList()
    protected var mMinScale = 1f
    protected var mMaxScale = 3f
    protected var mEdgeTapZone = 15f
    protected var mBackgroundColor = Color.DKGRAY

    // MARK: - Drawing

    protected val drawingController = DrawingController()
    protected val textAnnotationHandler: TextAnnotationHandler

    // MARK: - PDF State

    protected var mPdfRenderer: PdfRenderer? = null
    protected var mFileDescriptor: ParcelFileDescriptor? = null
    protected var mPdfPageWidth = 0
    protected var mPdfPageHeight = 0
    protected var mActualPageCount = 0

    // MARK: - Cache & Coroutines

    protected val mImageCache: LruCache<Int, Bitmap>
    protected val renderScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // MARK: - Page Tracking

    protected var mCurrentPage = 0
    protected var mPreviousWidth = 0

    init {
        drawingController.delegate = this
        textAnnotationHandler = TextAnnotationHandler(context)

        val maxMemory = (Runtime.getRuntime().maxMemory() / 1024).toInt()
        val cacheSize = maxMemory / 8
        mImageCache = object : LruCache<Int, Bitmap>(cacheSize) {
            override fun sizeOf(key: Int, bitmap: Bitmap): Int = bitmap.byteCount / 1024
        }
    }

    // MARK: - Override Points

    /** Called after PDF loaded successfully. Subclass should notify adapter and update UI. */
    protected abstract fun onPdfLoaded()

    /** Called after annotations changed. Subclass should refresh visible pages. */
    protected abstract fun onAnnotationsChanged()

    /** Called after drawing mode changed. Subclass should update gestures/input. */
    protected abstract fun onDrawingModeChanged(mode: DrawingMode)

    /** Called when drawing overlay needs redraw. */
    protected abstract fun redrawOverlay()

    // MARK: - Setters

    open fun setSource(source: String) {
        if (mSource != source) {
            mSource = source
            onSourceChanged()
        }
    }

    /** Override for deferred loading (e.g. PagingPdfView waits for valid dimensions). */
    protected open fun onSourceChanged() {
        reloadPdf()
    }

    fun setAnnotations(annotations: String?) {
        val newAnnotations = annotations ?: ""
        if (newAnnotations == mAnnotations) return
        mAnnotations = newAnnotations
        mParsedAnnotations = parseAnnotations(mAnnotations)
        mImageCache.evictAll()
        onAnnotationsChanged()
    }

    open fun setMinZoom(minZoom: Float) {
        mMinScale = minZoom.coerceAtLeast(0.5f)
    }

    open fun setMaxZoom(maxZoom: Float) {
        mMaxScale = maxZoom.coerceAtLeast(1f)
    }

    fun setEdgeTapZone(zone: Float) {
        mEdgeTapZone = zone.coerceIn(0f, 50f)
    }

    open fun setPdfBackgroundColor(color: Int) {
        if (color == mBackgroundColor) return
        mBackgroundColor = color
        onBackgroundColorChanged()
    }

    protected open fun onBackgroundColorChanged() {
        setBackgroundColor(mBackgroundColor)
    }

    fun setDrawingMode(mode: String) {
        val newMode = DrawingMode.fromString(mode)
        if (newMode == drawingController.drawingMode) return
        if (newMode != DrawingMode.TEXT) {
            textAnnotationHandler.commitTextInput()
        }
        drawingController.drawingMode = newMode
        onDrawingModeChanged(newMode)
    }

    fun setStrokeColor(color: String) {
        drawingController.strokeColor = color
    }

    fun setStrokeWidth(width: Float) {
        drawingController.strokeWidth = width
    }

    fun setStrokeOpacity(opacity: Float) {
        drawingController.strokeOpacity = opacity
    }

    fun setTextColor(color: String) {
        drawingController.textColor = color
        textAnnotationHandler.textColor = color
    }

    fun setTextFontSize(size: Float) {
        drawingController.textFontSize = size
        textAnnotationHandler.textFontSize = size
    }

    // MARK: - Public Commands

    abstract fun resetZoom()

    abstract fun scrollToPage(page: Int, animated: Boolean)

    fun clearStrokes(page: Int) {
        if (page < 0) {
            drawingController.clearAllStrokes()
            drawingController.clearAllTexts()
        } else {
            drawingController.clearStrokes(page)
            drawingController.clearTexts(page)
        }
        redrawOverlay()
    }

    fun getAnnotations(): WritableMap {
        return drawingController.getAnnotationsForExport()
    }

    // MARK: - PDF Loading

    protected fun reloadPdf() {
        if (mSource.isEmpty()) return

        closePdf()
        mImageCache.evictAll()

        val file = File(mSource)
        try {
            mFileDescriptor = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
            mPdfRenderer = PdfRenderer(mFileDescriptor!!)

            mPdfRenderer?.let { renderer ->
                if (renderer.pageCount > 0) {
                    val page = renderer.openPage(0)
                    mPdfPageWidth = page.width
                    mPdfPageHeight = page.height
                    page.close()
                }
                mActualPageCount = renderer.pageCount
            }

            onPdfLoaded()
            onLoadComplete()

        } catch (e: FileNotFoundException) {
            onError("File '$mSource' not found.")
        } catch (e: Exception) {
            onError("Failed to open PDF: ${e.message}")
        }
    }

    protected fun closePdf() {
        pdfMutex.withLock {
            try {
                mPdfRenderer?.close()
            } catch (_: Exception) {}
            mPdfRenderer = null
        }
        try {
            mFileDescriptor?.close()
        } catch (_: Exception) {}
        mFileDescriptor = null
    }

    protected fun getPageHeight(): Int {
        if (mPdfPageWidth <= 0 || mPdfPageHeight <= 0) return height
        return (width.toFloat() * mPdfPageHeight / mPdfPageWidth).toInt()
    }

    // MARK: - Cleanup

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        renderScope.cancel()
        closePdf()
        mImageCache.evictAll()
    }

    // MARK: - RN Event Emission

    protected fun emitEvent(eventName: String, event: WritableMap = Arguments.createMap()) {
        val reactContext = context as ReactContext
        reactContext.getJSModule(RCTEventEmitter::class.java).receiveEvent(id, eventName, event)
    }

    protected fun onError(message: String) {
        emitEvent("onPdfError", Arguments.createMap().apply { putString("message", message) })
    }

    protected fun onLoadComplete() {
        emitEvent("onPdfLoadComplete", Arguments.createMap().apply {
            putInt("width", mPdfPageWidth)
            putInt("height", mPdfPageHeight)
            putInt("pageCount", mActualPageCount)
        })
    }

    protected fun onPageChange() {
        emitEvent("onPageChange", Arguments.createMap().apply { putInt("page", mCurrentPage) })
    }

    protected fun onZoomChange(scale: Float) {
        emitEvent("onZoomChange", Arguments.createMap().apply { putDouble("scale", scale.toDouble()) })
    }

    protected fun onTap(position: String) {
        emitEvent("onTap", Arguments.createMap().apply { putString("position", position) })
    }

    protected fun onMiddleClick() {
        emitEvent("onMiddleClick")
    }

    // MARK: - DrawingControllerDelegate

    override fun onDrawingStart() {
        emitEvent("onDrawingStart")
    }

    override fun onDrawingEnd() {
        emitEvent("onDrawingEnd")
    }

    override fun onStrokeAdded(stroke: DrawingStroke, page: Int) {}

    override fun onStrokeRemoved(strokeId: String, page: Int) {}

    override fun onStrokesCleared(page: Int) {}

    override fun onNeedsRedraw() {
        post { redrawOverlay() }
    }
}

File: ios/PdfViewerBase.swift
import UIKit

// MARK: - Tap Zone Helper

enum TapZone {
    case left, middle, right
}

func classifyTapZone(point: CGPoint, in bounds: CGRect, edgeTapZone: CGFloat) -> TapZone {
    let edgeRatio = edgeTapZone / 100.0
    let leftEdge = bounds.width * edgeRatio
    let rightEdge = bounds.width * (1.0 - edgeRatio)
    if point.x < leftEdge { return .left }
    if point.x > rightEdge { return .right }
    return .middle
}

// MARK: - Content Inset Helper

func centeredContentInset(for scrollView: UIScrollView, extraTop: CGFloat = 0, extraBottom: CGFloat = 0) -> UIEdgeInsets {
    let scrollViewSize = scrollView.bounds.size
    let contentSize = scrollView.contentSize
    let scale = scrollView.zoomScale

    let scaledContentWidth = contentSize.width * scale
    let scaledContentHeight = contentSize.height * scale

    let horizontalInset = max(0, (scrollViewSize.width - scaledContentWidth) / 2)
    let verticalInset = max(0, (scrollViewSize.height - scaledContentHeight) / 2)

    return UIEdgeInsets(
        top: verticalInset + extraTop,
        left: horizontalInset,
        bottom: verticalInset + extraBottom,
        right: horizontalInset
    )
}

// MARK: - PdfViewerBase

/// Base class for PagingPdfView and ZoomablePdfScrollView.
/// Contains shared React props, PDF loading, annotation parsing, drawing controller, and stroke management.
@objc(PdfViewerBase) @objcMembers
class PdfViewerBase: UIView, DrawingControllerDelegate {

    // MARK: - React Props

    @objc var source = "" { didSet { reloadPdf() } }

    @objc var annotations = "" {
        didSet {
            parseAnnotations()
            imageCache.removeAllObjects()
            onAnnotationsChanged()
        }
    }

    @objc var minZoom: CGFloat = 1.0 { didSet { updateZoomLimits() } }
    @objc var maxZoom: CGFloat = 3.0 { didSet { updateZoomLimits() } }
    @objc var edgeTapZone: CGFloat = 15.0

    @objc var pdfBackgroundColor: UIColor = UIColor(white: 0.2, alpha: 1.0) {
        didSet { updateBackgroundColor() }
    }

    // MARK: - Drawing Props

    @objc var drawingMode = "view" { didSet { applyDrawingMode() } }
    @objc var strokeColor = "#000000" { didSet { drawingController.strokeColor = strokeColor } }
    @objc var strokeWidth: CGFloat = 3.0 { didSet { drawingController.strokeWidth = strokeWidth } }
    @objc var strokeOpacity: CGFloat = 1.0 { didSet { drawingController.strokeOpacity = strokeOpacity } }
    @objc var strokes = "" { didSet { loadStrokes() } }

    // MARK: - React Events

    @objc var onPdfError: RCTDirectEventBlock?
    @objc var onPdfLoadComplete: RCTDirectEventBlock? {
        didSet {
            if let pending = pendingLoadCompleteEvent {
                onPdfLoadComplete?(pending)
                pendingLoadCompleteEvent = nil
            }
        }
    }
    @objc var onPageChange: RCTDirectEventBlock?
    @objc var onZoomChange: RCTDirectEventBlock?
    @objc var onTap: RCTDirectEventBlock?
    @objc var onMiddleClick: RCTDirectEventBlock?
    @objc var onDrawingStart: RCTDirectEventBlock?
    @objc var onDrawingEnd: RCTDirectEventBlock?
    @objc var onUndoStateChange: RCTDirectEventBlock?

    // MARK: - Text Annotation Props

    @objc var textColor = "#0000FF" { didSet { textAnnotationHandler.textColor = textColor } }
    @objc var textFontSize: CGFloat = 16.0 { didSet { textAnnotationHandler.textFontSize = textFontSize } }

    // MARK: - Internal State (accessible to subclasses)

    private var pendingLoadCompleteEvent: [String: Any]?
    private(set) var pdfDocument: CGPDFDocument?
    var currentPage: Int = 0
    private var isReloading = false
    private(set) var actualPageCount: Int = 0

    private(set) var pdfPageWidth: CGFloat = 0
    private(set) var pdfPageHeight: CGFloat = 0

    let imageCache = NSCache<NSNumber, UIImage>()
    private(set) var parsedAnnotations: [AnnotationPage] = []

    let drawingController = DrawingController()
    var realDrawingMode = DrawingMode.view
    let textAnnotationHandler = TextAnnotationHandler()

    var previousBoundsWidth: CGFloat = 0

    // MARK: - Initialization

    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit()
    }

    private func commonInit() {
        backgroundColor = pdfBackgroundColor
        drawingController.delegate = self
        setupViews()
    }

    // MARK: - Override Points (subclasses must call super where noted)

    /// Setup viewer-specific views. Called once during init.
    /// Subclasses should set imageCache.countLimit here.
    func setupViews() {
        // Override in subclass
    }

    /// Called after PDF document loaded and dimensions extracted.
    func onPdfLoaded() {
        // Override in subclass
    }

    /// Called after annotations prop changed and parsed.
    func onAnnotationsChanged() {
        // Override in subclass
    }

    /// Called when minZoom/maxZoom changes.
    func updateZoomLimits() {
        // Override in subclass
    }

    /// Called when pdfBackgroundColor changes.
    func updateBackgroundColor() {
        backgroundColor = pdfBackgroundColor
    }

    /// Called after common drawing mode logic (realDrawingMode set, drawingController updated,
    /// text input dismissed). Subclass should update gestures/scroll here.
    func onDrawingModeChanged(_ mode: DrawingMode) {
        // Override in subclass
    }

    /// Called when the drawing overlay needs a redraw.
    func redrawCurrentOverlay() {
        // Override in subclass
    }

    // MARK: - PDF Loading

    func reloadPdf() {
        guard !source.isEmpty, !isReloading else { return }
        isReloading = true

        imageCache.removeAllObjects()
        pdfDocument = nil

        let url = URL(fileURLWithPath: source)
        guard let document = CGPDFDocument(url as CFURL) else {
            onPdfError?(["message": "Failed to open PDF: \(source)"])
            isReloading = false
            return
        }

        pdfDocument = document

        if let firstPage = document.page(at: 1) {
            let dims = firstPage.effectiveDimensions
            pdfPageWidth = dims.width
            pdfPageHeight = dims.height
        }

        actualPageCount = document.numberOfPages
        isReloading = false

        onPdfLoaded()

        let loadCompleteEvent: [String: Any] = [
            "width": pdfPageWidth,
            "height": pdfPageHeight,
            "pageCount": actualPageCount
        ]
        if let callback = onPdfLoadComplete {
            callback(loadCompleteEvent)
        } else {
            pendingLoadCompleteEvent = loadCompleteEvent
        }
    }

    // MARK: - Annotations

    private func parseAnnotations() {
        guard !annotations.isEmpty,
              let data = annotations.data(using: .utf8) else {
            parsedAnnotations = []
            return
        }

        do {
            parsedAnnotations = try JSONDecoder().decode([AnnotationPage].self, from: data)
        } catch {
            parsedAnnotations = []
        }
    }

    // MARK: - PDF Rendering

    func renderPage(at index: Int, completion: @escaping (UIImage?) -> Void) {
        guard let document = pdfDocument else {
            completion(nil)
            return
        }

        let annotation = index < parsedAnnotations.count ? parsedAnnotations[index] : nil

        PdfPageRenderer.renderPage(
            document: document,
            pageIndex: index,
            viewWidth: bounds.width,
            pdfPageWidth: pdfPageWidth,
            pdfPageHeight: pdfPageHeight,
            annotation: annotation,
            completion: completion
        )
    }

    // MARK: - Drawing Mode

    private func applyDrawingMode() {
        guard let mode = DrawingMode(rawValue: drawingMode) else {
            realDrawingMode = .view
            drawingController.drawingMode = .view
            return
        }
        realDrawingMode = mode
        drawingController.drawingMode = mode

        if mode != .text {
            textAnnotationHandler.commitTextInput()
        }

        onDrawingModeChanged(mode)
    }

    // MARK: - Strokes

    private func loadStrokes() {
        guard !strokes.isEmpty else {
            drawingController.clearAllStrokes()
            redrawCurrentOverlay()
            return
        }

        do {
            let data = strokes.data(using: .utf8)!
            let pageStrokes = try JSONDecoder().decode(PageStrokes.self, from: data)
            drawingController.setAllStrokes(pageStrokes)
            redrawCurrentOverlay()
        } catch {
            onPdfError?(["message": "Failed to parse strokes: \(error.localizedDescription)"])
        }
    }

    // MARK: - Public Commands

    func resetZoom() {
        // Override in subclass
    }

    func scrollToPage(_ page: Int, animated: Bool) {
        // Override in subclass
    }

    func undo() {
        drawingController.undo()
    }

    func redo() {
        drawingController.redo()
    }

    func clearStrokes(page: Int) {
        if page >= 0 {
            drawingController.clearStrokes(forPage: page)
            drawingController.clearTexts(forPage: page)
        } else {
            drawingController.clearAllStrokes()
            drawingController.clearAllTexts()
        }
        redrawCurrentOverlay()
    }

    func getAnnotations() -> [String: Any] {
        return drawingController.getAnnotationsForExport()
    }

    func loadAnnotations(_ json: String) {
        guard !json.isEmpty,
              let data = json.data(using: .utf8) else {
            print("[loadAnnotations] Empty or invalid JSON")
            return
        }

        do {
            let pages = try JSONDecoder().decode([AnnotationPage].self, from: data)
            print("[loadAnnotations] Parsed \(pages.count) pages")
            drawingController.clearAllStrokes()
            drawingController.clearAllTexts()

            for (pageIndex, page) in pages.enumerated() {
                let drawingStrokes = page.strokes.map { stroke in
                    DrawingStroke(
                        id: UUID().uuidString,
                        color: stroke.color,
                        width: stroke.width,
                        opacity: stroke.opacity ?? 1.0,
                        path: stroke.path
                    )
                }
                drawingController.setStrokes(drawingStrokes, forPage: pageIndex)

                for text in page.text {
                    guard text.point.count >= 2 else { continue }
                    let drawingText = DrawingText(
                        id: UUID().uuidString,
                        color: text.color,
                        fontSize: text.fontSize,
                        point: text.point,
                        str: text.str
                    )
                    drawingController.addText(drawingText, toPage: pageIndex)
                }

                if !page.strokes.isEmpty || !page.text.isEmpty {
                    print("[loadAnnotations] Page \(pageIndex): \(drawingStrokes.count) strokes, \(page.text.count) texts")
                    if let first = drawingStrokes.first, let firstPoint = first.path.first {
                        print("[loadAnnotations]   first stroke path[0]=\(firstPoint)")
                    }
                }
            }
            drawingController.clearUndoStack()
            redrawCurrentOverlay()
            print("[loadAnnotations] Done, overlay redrawn")
        } catch {
            print("[loadAnnotations] Parse error: \(error)")
            onPdfError?(["message": "Failed to parse loadAnnotations: \(error.localizedDescription)"])
        }
    }

    func clearCache() {
        imageCache.removeAllObjects()
    }

    // MARK: - DrawingControllerDelegate

    func drawingController(_ controller: DrawingController, didAddStroke stroke: DrawingStroke, onPage page: Int) {}

    func drawingController(_ controller: DrawingController, didRemoveStroke strokeId: String, onPage page: Int) {}

    func drawingController(_ controller: DrawingController, strokesCleared onPage: Int) {}

    func drawingControllerDidStartDrawing(_ controller: DrawingController) {
        onDrawingStart?([:])
    }

    func drawingControllerDidEndDrawing(_ controller: DrawingController) {
        onDrawingEnd?([:])
    }

    func drawingControllerNeedsRedraw(_ controller: DrawingController) {
        redrawCurrentOverlay()
    }

    func drawingController(_ controller: DrawingController, didRequestTextInputAt normalizedPoint: CGPoint, onPage page: Int) {
        // Handled by TextAnnotationHandler via touch interception
    }

    func drawingController(_ controller: DrawingController, undoStateChanged canUndo: Bool, canRedo: Bool) {
        onUndoStateChange?(["canUndo": canUndo, "canRedo": canRedo])
    }
}
